src/robot/variables/replacer.py

```diff
@@ -137,46 +137,59 @@ def _get_variable_item(self, match, value):
-                    % (name, type_name(value), item)
+                    "Variable '%s' is %s, which is not subscriptable, and "
+                    "thus accessing item '%s' from it is not possible. To use "
+                    "'[%s]' as a literal value, it needs to be escaped like "
+                    "'\\[%s]'." % (name, type_name(value), item, item, item)
                 )
             name = '%s[%s]' % (name, item)
         return value
 
-    def _get_list_variable_item(self, name, variable, index):
+    def _get_sequence_variable_item(self, name, variable, index):
         index = self.replace_string(index)
         try:
-            index = self._parse_list_variable_index(index, name[0] == '$')
+            index = self._parse_sequence_variable_index(index, name[0] == '$')
         except ValueError:
-            raise VariableError("List '%s' used with invalid index '%s'."
-                                % (name, index))
+            raise VariableError("Sequence '%s' used with invalid index '%s'. "
+                                "To use '[%s]' as a literal value, it needs "
+                                "to be escaped like '\\[%s]'."
+                                % (name, index, index, index))
         try:
             return variable[index]
         except IndexError:
-            raise VariableError("List '%s' has no item in index %d."
+            raise VariableError("Sequence '%s' has no item in index %d."
                                 % (name, index))
 
-    def _parse_list_variable_index(self, index, support_slice=True):
+    def _parse_sequence_variable_index(self, index, support_slice=True):
         if ':' not in index:
             return int(index)
         if index.count(':') > 2 or not support_slice:
             raise ValueError
         return slice(*[int(i) if i else None for i in index.split(':')])
 
-    def _get_dict_variable_item(self, name, variable, key):
-        key = self.replace_scalar(key)
+    def _get_subscriptable_variable_item(self, name, variable, key):
+        if not isinstance(key, (int, slice)):
+            key = self.replace_scalar(key)
         try:
             return variable[key]
         except KeyError:
-            raise VariableError("Dictionary '%s' has no key '%s'."
+            raise VariableError("Subscriptable '%s' has no key '%s'."
                                 % (name, key))
         except TypeError as err:
-            raise VariableError("Dictionary '%s' used with invalid key: %s"
+            if not isinstance(key, (int, slice)):
+                try:
+                    key = self._parse_sequence_variable_index(
+                        key, name[0] == '$')
+                except:
+                    pass
+                else:
+                    return self._get_subscriptable_variable_item(
+                        name, variable, key)
+            raise VariableError("Subscriptable '%s' used with invalid key: %s"
```

Comment:
Why is this done? Is the idea to automatically convert strings like "2" to an integers and "2:" to a slice? If yes, I don't think that's a good idea. We already do that with sequences and don't see much benefits with other objects. It's already now possible to do e.g. `${dict}[${2}]` to use non-string keys. That works also with other types than integers such as `${dict}[${1.2}]` and `${dict}[${True}]` while this code only supports integers and slices.

If there is a good use case this feature, or I have understood the code wrong, then  the plain `except:` should still be replaced with something that doesn't blindly swallow all exceptions.

Reply:
The above said, error handling would need to be changed a bit. This is because with custom objects `variable[key]` can fail with _any_ exception, not only with `TypeError`. Perhaps something like this:

```python
except:
    raise VariableError("Accessing item '%s' from %s '%s' failed: %s"
                        % (key, type_name(variable), name, get_error_message())
```

Notice that the plain `except:` is fine here because `get_error_message()` automatically re-raises exceptions such as `SystemExit` that we shouldn't catch.

Reply:
The idea was to support integers, slices and non-string keys for custom subscriptable classes, which is what this implementation is capable of. I now understand that you don't want this. I have simplified this implementation according to your suggestions. 2afca766be3f4a2058f4225895890babe4999c31

Reply:
I have simplified the implementation by using your suggestions in 2afca766be3f4a2058f4225895890babe4999c31. Updated documentation in 4ef6eb43757b0bfc0c38c31535a8504fa62c266b.

Reply:
The tests no longer pass after this simplified implementation. See explanation in https://github.com/robotframework/robotframework/pull/3350#discussion_r386965038

Reply:
I refactored the first implementation in d7110621c16a371e72b541b70f01cbea7f0f4c8e. Existing tests pass and I added some additional unit tests.